### DISCUSSION TITLE
Hyphenation edge cases

### DISCUSSION THREAD

**Discussion by biggnome** [sile-typesetter/sile#1377] [Q&A] [answered]
I was working on a document the other night, and ran into a couple stubborn hyphenation issues. One of these was "whirlpool" not breaking and spilling out of frame (I could imagine other compound words manifesting the same problem, although I tested a few out--e.g., "breastplate", "boardroom", etc.--and those all break correctly); the other was "o’clock" breaking with a hyphen after the "o’" (I tested a few similar words, which were hit-and-miss, e.g., "jack o’lantern" and "O’Leary" are okay, but "O’Malley" hyphenates in the same questionable way "o’clock" does).

Initially, I worked around this by adding "whirl-pool" into the exceptions array in en.lua and enclosed a couple of the "o’clocks" in hboxes, although that's a pretty dirty kludge for obvious reasons. When I came back to the project after work this afternoon, I scrapped that and instead added the following to the top of my document (I'm afraid my programming experience is very limited and I don't really "speak" Lua, and I'll bet it shows)--

```
\language[main=en]
\begin{script}
    local hyphy = {"whirl-pool", "o’clock"}

    for i=0, #hyphy do
      table.insert(SILE.hyphenator.languages["en"].exceptions, hyphy[i])
    end
\end{script}
```

--which seems to work just as well and certainly scales a lot better than manually editing en.lua and/or using hboxes. Is there a more elegant (or just less janky...) way to approach this sort of thing? It seems inapt to me to be invoking Lua outside of a class, at least for something so trivial.

**Comment by alerque** [comment, marked as answer]
First, your Lua doesn't look half bad, don't sweat it. It tends to be a bit verbose and can look clunky — a price you pay for speed in an interpreted language.

<details>
<summary>If you did want a slightly more ergonomic Lua loop:</summary>

Not that this is the best way to approach the hyphenation issue, but just looking at the Lua look you can use `ipairs()` for your example like this:

```lua
local hyphy = {"whirl-pool", "o’clock"}
for _, exception in ipairs(hyphy) do
	table.insert(SILE.hyphenator.languages["en"].exceptions, exception)
end
```

Not much difference but sometimes easier to read. The `_` variable I discarded here could be `i` if you need the key for anything in the loop.

</details>

Second, don't ever worry about adding Lua outside of a class. You may consider doing it in a package/library of your own rather than in a document, but being able to use Lua whenever/wherever you need it.

On the other hand we did build in a more ergonomic way to add exceptions. The API we added though was for the SILE language itself, not necessarily for easy Lua access. You can see the code in `core/hyphenator-liang.lua`. From the SILE input language for yoru example you could call:

```sile
\begin[lang=en]{hyphenator:add-exceptions}
whirl-pool
o’clock
\end{hyphenator:add-exceptions}
```

You can of course also call this function from Lua:

```lua
SILE.call(hyphenator:add-exceptions, { lang = "en" }, {[[
whirl-pool
o’clock
]]})
```

**Comment by alerque** [comment]
Also note English isn't the only language that has issues with a few words. See for example [this list of mostly proper names with apostrophes](https://github.com/sile-typesetter/casile/blob/master/hyphenation_exceptions.lua) that I manually added to CaSILE to handle common problem spots in Turkish books I've publish.

**Comment by alerque** [comment]
Just another note if you are having trouble with frame overflows, you might consider tweaking some other properties. I for one almost always change SILE's defaults to be a bit more forgiving of hard to typeset content. For example some things I do in Turkish on a regular basis:

```lua
SILE.settings.set("linebreak.emergencyStretch", SILE.length.parse("3spc"))
SILE.settings.set("linebreak.finalHyphenDemerits", 500)
SILE.settings.set("linebreak.hyphenPenalty", 100)
SILE.settings.set("typesetter.overfulltolerance", SILE.length("0.2ex"))
SILE.settings.set("typesetter.underfulltolerance", SILE.length("6ex"))
SILE.settings.set("document.spaceskip", SILE.length.parse("1spc plus 2spc minus 0.3spc"))
```

I'm not saying these values are the best for all projects. I even use different values on different projects working my way up from the defaults until things work well. These are arguably *bad* typography as it can result in huge spaces in lines, but sometimes that's better than things like URLs or large strings with no breaks overflowing the frame.

**Comment by biggnome** [comment]
Wow! Thanks for the insight (and the kind words). I'd totally missed `hyphenator:add-exceptions`, and while I mostly have (otherwise--hence why I raised the question) found SILE to be pretty well-behaved, I'll definitely keep the linebreaking, etc. settings you mentioned in mind as well--those I was aware of, but have not yet encountered a practical need to adjust them.

**Comment by Omikhleia** [comment]
In the TeX/hunspell hyphenation patterns, whirlpool has no hyphenation in en-US but has one in en-GB. I'm not sure which one SILE uses (but that could be an interesting test case for supporting language subtags)...

The "o'clock" case might be a bug, however.

**Reply by simoncozens** [reply]
It is; we’re treating punctuation as part of the word for hyphenation purposes, which is wrong.

**Reply by alerque** [reply]
Wrong in some languages, not others. And sometimes even when they should not be considered as part of the word for hyphenation you have to drop them if you do hyphenate, ala #355. Fun times.
